File: Algorithms/src/1878. Get Biggest Three Rhombus Sums in a Grid/Code/solution.js
var getBiggestThree = function (grid) {
  let m = grid.length;
  let n = grid[0].length;
  let set = new Set();

  for (let r = 0; r < m; r++) {
    for (let c = 0; c < n; c++) {
      set.add(grid[r][c]);

      let maxSize = Math.min(r, c, m - 1 - r, n - 1 - c);

      for (let k = 1; k <= maxSize; k++) {
        let sum = 0;

        for (let i = 0; i < k; i++) sum += grid[r - k + i][c + i];

        for (let i = 0; i < k; i++) sum += grid[r + i][c + k - i];

        for (let i = 0; i < k; i++) sum += grid[r + k - i][c - i];

        for (let i = 0; i < k; i++) sum += grid[r - i][c - k + i];

        set.add(sum);
      }
    }
  }

  let res = Array.from(set);
  res.sort((a, b) => b - a);

  return res.slice(0, 3);
};

File: Algorithms/src/1878. Get Biggest Three Rhombus Sums in a Grid/Code/solution.cpp
class Solution
{
public:
    vector<int> getBiggestThree(vector<vector<int>> &grid)
    {
        int m = grid.size();
        int n = grid[0].size();

        set<int> sums;

        for (int r = 0; r < m; r++)
        {
            for (int c = 0; c < n; c++)
            {

                // size 0 rhombus (single cell)
                sums.insert(grid[r][c]);

                int maxSize = min({r, c, m - 1 - r, n - 1 - c});

                for (int k = 1; k <= maxSize; k++)
                {
                    int sum = 0;

                    int x = r - k, y = c;

                    // top -> right
                    for (int i = 0; i < k; i++)
                    {
                        sum += grid[x + i][y + i];
                    }

                    // right -> bottom
                    for (int i = 0; i < k; i++)
                    {
                        sum += grid[r + i][c + k - i];
                    }

                    // bottom -> left
                    for (int i = 0; i < k; i++)
                    {
                        sum += grid[r + k - i][c - i];
                    }

                    // left -> top
                    for (int i = 0; i < k; i++)
                    {
                        sum += grid[r - i][c - k + i];
                    }

                    sums.insert(sum);
                }
            }
        }

        vector<int> res(sums.begin(), sums.end());
        sort(res.rbegin(), res.rend());

        if (res.size() > 3)
            res.resize(3);

        return res;
    }
};